### PR TITLE
fix JSON schema with MLM fields + support pydantic/pystac objects

### PR DESCRIPTION
**Pull request by fmigneault**
PR just to demonstrate the changes from https://github.com/crim-ca/dlm-extension/pull/2

### CODE REVIEW

**Comment by rbavery** [comment]
@fmigneault awesome thanks. I can review this next week when it is ready.

**Comment by fmigneault** [comment]
@rbavery 
I think I have addressed most of the editorial items. 
I'll work on the JSON-schema definition to reflect the descriptions at the start of next week.

**Comment by fmigneault** [comment]
@rbavery Almost there with the JSON schema. Only a few definitions about `mlm:output` and the MLM-roles left to implement. I did a lot of adjustments to the README to place the sections closer to where they can mention first. There was a lot of back and forth between Item/Asset fields when following links.

File: json-schema/schema.json [hunk 1]
```diff
@@ -1,557 +1,793 @@
 {
   "$schema": "http://json-schema.org/draft-07/schema#",
-  "$id": "https://schemas.stacspec.org/v1.0.0-beta.3/extensions/dl-model/json-schema/schema.json",
+  "$id": "https://stac-extensions.github.io/mlm/v1.0.0/schema.json",
   "title": "DL Model Item",
```
Comment (devisperessutti) [inline json-schema/schema.json]: should this be updated as well to `ML Model Item` ?

File: README.md [hunk 2]
```diff
@@ -43,118 +55,524 @@ Check the original technical report for an earlier version of the Model Extensio
...
+| `classification`        | `classification`            | Generic classification task that assigns class labels to an output.                                             |
+| `scene-classification`  | *n/a*                       | Specific classification task where the model assigns a single class label to an entire scene/area.              |
+| `detection`             | `detection`                 | Generic detection of the "presence" of objects or entities, with or without positions.                          |
+| `object-detection`      | *n/a*                       | Task corresponding to the identification of positions as bounding boxes of object detected in the scene.        |
+| `segmentation`          | `segmentation`              | Generic tasks that regroups all types of segmentations tasks consisting of applying labels to pixels.           |
+| `semantic-segmentation` | *n/a*                       | Specific segmentation task where all pixels are attributed labels, without consideration of similar instances.  |
```
Comment (rbavery) [inline README.md:134]:
```suggestion
| `semantic-segmentation` | *n/a*                       | Specific segmentation task where all pixels are attributed labels, without consideration for segments as unique objects.  |
```

File: README.md [hunk 3]
```diff
@@ -43,118 +55,524 @@ Check the original technical report for an earlier version of the Model Extensio
...
+It is to be noted that multiple "*generic*" tasks names (`classification`, `detection`, etc.) are defined to allow
+correspondance with `label:tasks`, but these can lead to some ambiguity depending on context. For example, a model
+that supports `classification` could mean that the model can predict patch-based classes over an entire scene
+(i.e.: `scene-classification` for a single prediction over an entire area of interest as a whole),
+or that it can predict pixel-wise "classifications", such as land-cover labels for
+every single pixel coordinate over the area of interest. Maybe counter-intuitively to some users,
+such a model that produces pixel-wise "classifications" should be attributed the `segmentation` task
+(and more specifically `semantic-segmentation`) rather than `classification`. To avoid this kind of ambiguity,
+it is strongly recommended that `tasks` always aim to provide the most specific definitions possible to explicitly
+describe what the model accomplishes.
```
Comment (rbavery) [inline README.md:155]: great explanation here!

File: README.md [hunk 4]
```diff
@@ -43,118 +55,524 @@ Check the original technical report for an earlier version of the Model Extensio
...
+- `TensorFlow`
+- `Scikit-learn`
+- `Huggingface`
+- `PyMC`
+- `JAX`
+- `ONNX`
+- `MXNet`
+- `Keras`
+- `Caffe`
+- `Weka`
```
Comment (rbavery) [inline README.md:175]: should `Weka` be listed? I've never heard of it or seen it in the wild.

Here's a suggested reordering based on my subjective interpretation of current popularity + longevity. 

I also added `rgee` and `spatialRF` to showcase some R options. especially in academia, lots of folks use R, particularly random forest models for semantic segmentation.

I removed Caffe (no updates in 4 years) and MxNet (archived last year). I don't think anyone will publish models for these frameworks.

Removed ONNX since it isn't a training framework and I think the purpose of this field is to describe the framework used to train the model. this might be different than the inference runtime and format.

```suggestion
- `PyTorch`
- `TensorFlow`
- `Scikit-learn`
- `Huggingface`
- `Keras`
- `rgee`
- `spatialRF`
- `JAX`
- `PyMC`
```

File: README.md [hunk 5]
```diff
@@ -43,118 +55,524 @@ Check the original technical report for an earlier version of the Model Extensio
...
-| norm_with_clip_values   | [integer]                                                                        | If `norm_type = "norm_with_clip"` this array supplies a value that is less than the band maximum. The array must be the same length as "bands", each value is used to divide each band before clipping values between 0 and 1.                     |
-| pre_processing_function | string                                                                           | A url to the preprocessing function where normalization and rescaling takes place, and any other significant operations. Or, instead, the function code path, for example: `my_python_module_name:my_processing_function`                          |     |
+### Framework
 
-### Accelerator Enum
+In most cases, this should correspond to common library names of well-established ML frameworks.
```
Comment (rbavery) [inline README.md:161]:
```suggestion
This should correspond to the common library name of the well-established ML framework used to train the model.
```

File: README.md [hunk 6]
```diff
@@ -43,118 +55,524 @@ Check the original technical report for an earlier version of the Model Extensio
...
+- `interpolation-nearest`
+- `interpolation-linear`
+- `interpolation-cubic`
+- `interpolation-area`
+- `interpolation-lanczos4`
+- `interpolation-max`
+- `wrap-fill-outliers`
+- `wrap-inverse-map`
+
+See [OpenCV - Normalization Flags](https://docs.opencv.org/4.x/d2/de8/group__core__array.html#ga87eef7ee3970f86906d69a92cbf064bd)
```
Comment (rbavery) [inline README.md:348]: I think this reference to normalization flags needs to be switched with the reference to interpolation.

long term I'd be interested in picking a different reference than OpenCV's C++ documentation, since the OpenCV lib is lower level than most folks encounter and the docs  are a bit hard to follow (python programmer might get confused with the C data types for example). But I think this is better than us rolling our own.

File: README.md [hunk 7]
```diff
@@ -43,118 +55,524 @@ Check the original technical report for an earlier version of the Model Extensio
...
+In each case, the `mlm:model` should be applied as well to indicate that this asset represents the model.
+
+It is also recommended to make use of the
+[file](https://github.com/stac-extensions/file?tab=readme-ov-file#asset--link-object-fields)
+extension for this Asset, as it can provide useful information to validate the contents of the model definition,
+by comparison with fields `file:checksum` and `file:size` for example.
+
+#### Model Artifact Media-Type
+
+Not all ML framework, libraries or model artifacts provide explicit media-type. When those are not provided, custom
```
Comment (rbavery) [inline README.md:504]:
```suggestion
Not all ML framework, libraries or model artifacts provide an explicit media-type. When those are not provided, custom
```

File: README.md [hunk 8]
```diff
@@ -43,118 +55,524 @@ Check the original technical report for an earlier version of the Model Extensio
...
+media-types can be considered. For example `application/x-pytorch` or `application/octet-stream; application=pytorch`
+could be appropriate to represent a PyTorch `.pt` file, since the underlying format is a serialized pickle structure.
+
+#### Artifact Type Enum
+
+This value can be used to provide additional details about the specific model artifact being described.
+For example, PyTorch offers various strategies for providing model definitions, such as Pickle (`.pt`), TorchScript,
+or the compiled approach. Since they all refer to the same ML framework,
```
Comment (rbavery) [inline README.md:512]:
```suggestion
or the upcoming [Ahead-Of-Time Compiled .pt2 format](https://pytorch.org/docs/main/torch.compiler_aot_inductor.html). Since they all refer to the same ML framework,
```

File: README.md [hunk 9]
```diff
@@ -43,118 +55,524 @@ Check the original technical report for an earlier version of the Model Extensio
...
+the [Model Artifact Media-Type](#model-artifact-media-type) would be insufficient in this case to detect with strategy
+should be used. 
+
+Following are some proposed *Artifact Type* values for corresponding approaches, but other names are
+permitted as well. Note that the names are selected using the framework-specific definitions to help
+the users understand the source explicitly, although this is not strictly required either.
+
+| Artifact Type      | Description                                                                                                              |
+|--------------------|--------------------------------------------------------------------------------------------------------------------------|
+| `torch.compile`    | A model artifact obtained by [`torch.compile`](https://pytorch.org/tutorials/intermediate/torch_compile_tutorial.html).  |
```
Comment (rbavery) [inline README.md:522]:
```suggestion
| `.pt2`    | A model artifact obtained using [Pytorch's AOTInductor](https://pytorch.org/docs/main/torch.compiler_aot_inductor.html).  |
```

I think this is a necessary edit since torch.compile is an API for compiling pytorch nn.Modules. it's used to speed up Pytorch code in general, for training or inference. Many of the backend internals that make torch.compile work are used in AOTInductor (the tool that creates compiled model artifacts) but they aren't the same thing and I think here we want to refer to artifacts produced by AOTInductor

File: README.md [hunk 10]
```diff
@@ -43,118 +55,524 @@ Check the original technical report for an earlier version of the Model Extensio
...
+| `torch.jit.script` | A model artifact obtained by [`TorchScript`](https://pytorch.org/docs/stable/jit.html).                                  |
```
Comment (rbavery) [inline README.md:523]:
```suggestion
| `torchscript` | A model artifact obtained by [`TorchScript Scripting`](https://pytorch.org/docs/stable/jit.html) and/or [`TorchScript Tracing`](https://pytorch.org/docs/stable/generated/torch.jit.trace.html).                                  |
```

There are two types of graph capture in torchscript, trace and script. I think we can either enumerate both or leave only one option. Either traced or scripted models can be loaded the same way so I favor only one field for both of them.

Somewhat confusingly, both can also be used together though this is not common. https://ppwwyyxx.com/blog/2022/TorchScript-Tracing-vs-Scripting/

Since Torchscript is bring phased out in favor of AOTInductor, I think we shouldn't make this too complex and only provide them as one field.

File: stac_model/output.py [hunk 11]
```diff
@@ -1,41 +1,94 @@
...
+from stac_model.base import DataType, MLMBaseModel, ModelTask, ProcessingExpression, OmitIfNone
+
+
+class ModelResult(MLMBaseModel):
+    shape: List[Union[int, float]] = Field(..., min_items=1)
+    dim_order: List[str] = Field(..., min_items=1)
+    data_type: DataType
+
+
+# MLMClassification: TypeAlias = Annotated[
```
Comment (rbavery) [inline stac_model/output.py:16]: can be deleted

**Review by rbavery — CHANGES_REQUESTED** [review]
@fmigneault this looks great. Some requests to edit the documentation and revised schema
